File: ruby/iforevents/lib/iforevents/events.rb
# frozen_string_literal: true

module Iforevents
  # The app's own user id plus traits (context merged in, nested hashes flattened).
  IdentifyEvent = Struct.new(:custom_id, :traits, keyword_init: true) do
    def initialize(custom_id:, traits: {})
      super(custom_id: custom_id, traits: traits || {})
    end
  end

  # A tracked event ready for every integration. +type+ is "track" or "page_view".
  TrackEvent = Struct.new(:name, :type, :properties, :timestamp, keyword_init: true) do
    def initialize(name:, type: "track", properties: {}, timestamp: Time.now.utc)
      super(name: name, type: type.to_s.empty? ? "track" : type, properties: properties || {}, timestamp: timestamp)
    end
  end

  # A page (web) or screen (mobile) view.
  PageEvent = Struct.new(:name, :properties, :navigation_type, :to_route, :previous_route, :timestamp, keyword_init: true) do
    def initialize(name: "page_view", properties: {}, navigation_type: nil, to_route: nil, previous_route: nil, timestamp: Time.now.utc)
      super(name: name.to_s.empty? ? "page_view" : name, properties: properties || {}, navigation_type: navigation_type, to_route: to_route, previous_route: previous_route, timestamp: timestamp)
    end
  end

  # Outcome of one integration call; the facade never raises for these.
  IntegrationResult = Struct.new(:integration, :success, :error, :timestamp, keyword_init: true) do
    def success?
      success
    end
  end

  # Flattens nested hashes with "_": {a: {b: 1}} becomes {"a_b" => 1}. Keys become strings.
  def self.flatten(hash, prefix = nil)
    out = {}
    (hash || {}).each do |key, value|
      name = prefix ? "#{prefix}_#{key}" : key.to_s
      if value.is_a?(Hash)
        out.merge!(flatten(value, name))
      else
        out[name] = value
      end
    end
    out
  end
end

File: go/events.go
// Package iforevents is the IForevents analytics SDK for Go: one facade,
// pluggable integrations, and a first-party API integration with batching,
// retries and typed errors. Only the public project key is ever configured;
// it grants event ingestion and nothing else.
package iforevents

import "time"

// Properties are free-form event or trait properties. Nested maps are
// flattened with "_" before reaching integrations.
type Properties map[string]any

// Event types the api distinguishes.
const (
	EventTypeTrack    = "track"
	EventTypePageView = "page_view"
)

// IdentifyEvent is the app's own user id plus traits (context merged in,
// nested maps flattened).
type IdentifyEvent struct {
	CustomID string
	Traits   Properties
}

// TrackEvent is a tracked event ready for every integration.
type TrackEvent struct {
	Name       string
	Type       string
	Properties Properties
	// Timestamp is when the event was queued; sent as created_at.
	Timestamp time.Time
}

// PageEvent is a page (web) or screen (mobile) view.
type PageEvent struct {
	Name           string
	Properties     Properties
	NavigationType string
	ToRoute        string
	PreviousRoute  string
	Timestamp      time.Time
}

// IntegrationResult is the outcome of one integration call; the facade never
// returns an error for these.
type IntegrationResult struct {
	Integration string
	Success     bool
	Err         error
	Timestamp   time.Time
}

// Flatten flattens nested maps with "_": {"a": {"b": 1}} becomes {"a_b": 1}.
// Slices and other values are kept as they are.
func Flatten(in Properties) Properties {
	out := Properties{}
	flattenInto(out, "", in)
	return out
}

func flattenInto(out Properties, prefix string, in map[string]any) {
	for k, v := range in {
		name := k
		if prefix != "" {
			name = prefix + "_" + k
		}
		switch nested := v.(type) {
		case Properties:
			flattenInto(out, name, nested)
		case map[string]any:
			flattenInto(out, name, nested)
		default:
			out[name] = v
		}
	}
}
